### PR TITLE
adding support for Italian Umberto ( umberto.commoncrawl and umberto.wikipedia ) from Musixmatch

### PR DESCRIPTION
**Pull request by simonefrancia**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes # (issue).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by simonefrancia** [comment]
Don't understand what the error could be from log.

**Comment by myleott** [comment]
Ah you can ignore the Windows build errors for now, the Ubuntu and MacOS builds passed.

Can you also add a readme similar to the Camembert one: https://github.com/pytorch/fairseq/blob/master/examples/camembert/README.md

**Comment by simonefrancia** [comment]
We would like to do the second option. 
Where do we have to move the model in `hubconf.py`?

**Comment by myleott** [comment]
I think a `hubconf.py` like this in your root directory should work:
```
dependencies = [
    'fairseq',
    'sentencepiece',
    'torch',
]

def umberto_commoncrawl_cased(**kwargs):
    from fairseq import hub_utils
    from fairseq.models.roberta.hub_interface import RobertaHubInterface
    x = hub_utils.from_pretrained(
        model_name_or_path='https://mxmdownloads.s3.amazonaws.com/umberto/umberto.commoncrawl.cased.tar.gz',
        checkpoint_file='model.pt',
        data_name_or_path='.',
        bpe='sentencepiece',
        load_checkpoint_heads=True,
        **kwargs,
    )
    return RobertaHubInterface(x['args'], x['task'], x['models'][0])
```

**Comment by simonefrancia** [comment]
Sorry but I cannot understand what you mean.
I already put this code:
```python
@register_model('umberto')
class UmbertoModel(RobertaModel):
    @classmethod
    def hub_models(cls):

        return {
            'umberto.wikipedia.uncased': 'https://mxmdownloads.s3.amazonaws.com/umberto/umberto.wikipedia.uncased.tar.gz',
            'umberto.commoncrawl.cased': 'https://mxmdownloads.s3.amazonaws.com/umberto/umberto.commoncrawl.cased.tar.gz'
        }

    @classmethod
    def from_pretrained(cls, model_name_or_path, checkpoint_file='model.pt', data_name_or_path='.', bpe='sentencepiece', **kwargs):
        from fairseq import hub_utils
        x = hub_utils.from_pretrained(
            model_name_or_path,
            checkpoint_file,
            data_name_or_path,
            archive_map=cls.hub_models(),
            bpe=bpe,
            load_checkpoint_heads=True,
            **kwargs,
        )
        
        return RobertaHubInterface(x['args'], x['task'], x['models'][0])
```

in `fairseq/fairseq/models/roberta/model.py `. In this way I can , as CamembertModel, download pretrained umberto from my own link.
So what should I do? This code is working and we are using it.

For second option do I have to write the code you provide me appending it to `fairseq/hubconf.py?`
I would like to remove musixmatch fork in order to have support for Umberto in your master Fairseq.


Thanks

**Comment by simonefrancia** [comment]
For us it would be better to be included as Camembert , here it's our README
```
# UmBERTo: an Italian Language Model trained with Whole Word Masking

## Introduction
[UmBERTo](https://github.com/musixmatchresearch/umberto) is a Roberta-based Language Model trained on large Italian Corpora. This implementation is based on Facebook Research AI code (https://github.com/pytorch/fairseq). It uses two innovative approaches: SentencePiece and Whole Word Masking.

Also available in [github.com/huggingface/transformers](https://huggingface.co/Musixmatch/umberto-commoncrawl-cased-v1)
## Pre-trained models

Model | #params | Vocab size | Corpus size | Download
---|---|---|---|---
`umberto.wikipedia.uncased` | 110M | 32k | ~7 GB | [umberto.wikipedia.uncased.tar.gz](https://mxmdownloads.s3.amazonaws.com/umberto/umberto.wikipedia.uncased.tar.gz)
`umberto.commoncrawl.cased` | 110M | 32k | ~70 GB | [umberto.commoncrawl.cased.tar.gz]( https://mxmdownloads.s3.amazonaws.com/umberto/umberto.commoncrawl.cased.tar.gz)

## Example usage

##### Load UmBERTo from torch.hub (PyTorch >= 1.1):
```python
#The model will be available soon
import torch
umberto = torch.hub.load('pytorch/fairseq', 'umberto.commoncrawl.cased')
umberto.eval()  # disable dropout (or leave in train mode to finetune)
```

##### Load UmBERTo (for PyTorch 1.0 or custom models):
```python
# Download umberto model
wget https://mxmdownloads.s3.amazonaws.com/umberto/umberto.commoncrawl.cased.tar.gz
tar -xzvf umberto.commoncrawl.cased.tar.gz

# Load the model in fairseq
from fairseq.models.roberta import UmbertoModel
umberto = UmbertoModel.from_pretrained('/path/to/umberto.commoncrawl.cased')
umberto.eval()  # disable dropout (or leave in train mode to finetune)
```

##### Filling masks:
```python
# Masked LM Inference
masked_line = 'Umberto Eco è <mask> un grande scrittore'
umberto.fill_mask(masked_line, topk=4)
# Output:
#('Umberto Eco è considerato un grande scrittore', 0.19939924776554108, ' considerato'), 
#('Umberto Eco è sicuramente un grande scrittore', 0.1669664829969406, ' sicuramente'), 
#('Umberto Eco è stato un grande scrittore', 0.16225320100784302, ' stato'), 
#('Umberto Eco è indubbiamente un grande scrittore', 0.09528309106826782, ' indubbiamente')
```

##### Extract features from UmBERTo:
```python
# Extract the last layer's features
line = "Il mio nome è Eco, Umberto Eco."
tokens = umberto.encode(line)
last_layer_features = umberto.extract_features(tokens)
assert last_layer_features.size() == torch.Size([1, 11, 768])

# Extract all layer's features (layer 0 is the embedding layer)
all_layers = umberto.extract_features(tokens, return_all_hiddens=True)
assert len(all_layers) == 13
assert torch.all(all_layers[-1] == last_layer_features)
```

**Comment by simonefrancia** [comment]
This could be our README

**Comment by myleott** [comment]
Since you're hosting the model yourself, I'm a bit reluctant to put this in fairseq, since (1) it introduces a dependency on servers that we don't control, and (2) if you release a new model it will require us to review PRs.

I'm suggesting that instead of putting this in fairseq, it will be easier to directly put a `hubconf.py` file in https://github.com/musixmatchresearch/umberto. That way people can do:
```
import torch
umberto = torch.hub.load('musixmatchresearch/umberto', 'umberto_commoncrawl_cased')
```

You don't need any fairseq changes to make the above work. Since you just use the standard RoBERTa interface, all you need to do is create a file named `hubconf.py` (with the code I pasted above) in the `github.com/musixmatchresearch/umberto` repo and it should work.

> I would like to remove musixmatch fork in order to have support for Umberto in your master Fairseq.

Yes, this is fine. Once you add the hubconf.py above you won't need to use a forked copy of fairseq.

**Comment by simonefrancia** [comment]
Ok thanks we inserted that file in our repo! Sorry for multiple questions. 
So in this way Umberto is available,  but could you please cite our repo in fairseq?
Thank you very much

**Comment by loretoparisi** [comment]
@myleott thank you very much. We are very glad to contribute to `fairseq`. By the way, what about integrate `umBERTo` in your models, so that there will be any external dependency? Of course if it makes sense for you. Thanks!

**Comment by myleott** [comment]
> could you please cite our repo in fairseq

Sure, I'll update this PR with a reference to the musixmatchresearch/umberto repo :)
